File: ch1/string_compression/python/strcomp.py
def stringComp(a):
    tmp = ""
    if(a == ""):
        return ""
    count = 1
    for i in range (len(a) - 1):
        if(a[i] == a[i + 1]):
            count += 1;
        else:
            tmp += a[i] + str(count)
            count = 1
    tmp += a[len(a) - 1] + str(count)
    return tmp

def main():
    a = "aabcccccaaa"
    b = "abc"
    c = "ggggggggggggggggggggggggggggggggggg"
    d = "aaabbbcdddddddddddddddddddddde"
    e = ""
    print(stringComp(a))
    print(stringComp(b))
    print(stringComp(c))
    print(stringComp(d))
    print(stringComp(e))

if __name__ == "__main__":
    main()

File: ch1/string_compression/java/StrComp.java
public class StrComp{
    public static String StrComp(String s){
        String tmp = "";
        if(s.isEmpty()){
            return "";
        }
        int count = 1;
        for(int i = 0; i < s.length() - 1; i++){
            if(s.charAt(i) == s.charAt(i + 1)){
                count++;
            } else {
                tmp += s.charAt(i) + String.valueOf(count);
                count = 1;
            }
        }
        tmp += s.charAt(s.length() - 1) + String.valueOf(count);
        return tmp;
    }
    
    public static void main(String[] args){
        String a = "abc";
        String b = "aabcccccaaa";
        String c = "";
        System.out.println(StrComp(c));
        System.out.println(StrComp(b));
        System.out.println(StrComp(a));
    }
}
